### PR TITLE
rgw: unlock on error in process_single_logshard()

### PR DESCRIPTION
**Pull request by weiqiaomiao**
Signed-off-by: Wei Qiaomiao <wei.qiaomiao@zte.com.cn>

### CODE REVIEW

**Comment by weiqiaomiao** [comment]
@joscollin  josconllin, it is no exception here，do you mean `try - finally` or can you give me an example?
Thanks very much！

**Comment by joscollin** [comment]
@weiqiaomiao 

I meant, instead of `goto` statement, you can just `throw` an integer and catch it at the place where you use `done:`. So the `unlock()` would be inside the `catch` block.

Search in the Ceph code for examples. A quick search shows me:
rgw_auth_keystone.cc -> L48.
rgw_auth.cc -> L210.

Even the caller of `process_single_logshard()` can also implement the same, but I would prefer implementing inside the function itself.

**Comment by joscollin** [comment]
@weiqiaomiao ping

**Comment by weiqiaomiao** [comment]
@joscollin sorry,I forgot that. I try to implement it with `try -- catch`, cound you take a look?

**Comment by weiqiaomiao** [comment]
> Could you please provide a detailed description (in the PR Description) about what you are trying to implement ? I mean, does this should unlock when catch an error is meant for all errors in process_single_logshard() ? It is not doing so here: https://github.com/weiqiaomiao/ceph/blob/4a1b98195d3ab2b59bff27d7e5f9faa813c5d7f0/src/rgw/rgw_reshard.cc#L839

  I not sure we need unlock here, if we catch `-EBUSY` here, it mean unanother processor already locked it.

> Could you please align your changes properly, so that it avoids the unwanted spaces highlighted (in the difference shown) ?
 
Sure!

**Comment by joscollin** [comment]
@weiqiaomiao Let me know when you are done with the changes.

**Comment by weiqiaomiao** [comment]
@joscollin done.many thanks!

**Comment by weiqiaomiao** [comment]
jenkins test this please

**Comment by weiqiaomiao** [comment]
@joscollin OK

File: src/rgw/rgw_reshard.cc [hunk 1]
```diff
@@ -832,13 +834,17 @@ int RGWReshard::process_single_logshard(int logshard_num)
         ret = l.lock_exclusive(&store->reshard_pool_ctx, logshard_oid);
         if (ret == -EBUSY) { /* already locked by another processor */
           ldout(store->ctx(), 5) << __func__ << "(): failed to acquire lock on " << logshard_oid << dendl;
-          return ret;
+          throw ret;
```
Comment (joscollin) [inline src/rgw/rgw_reshard.cc]: Looks better now. I think what you said before makes sense. Better don't unlock here, as we already got an `-EBUSY` when trying to lock.

Comment (weiqiaomiao) [reply src/rgw/rgw_reshard.cc]: agreed.  already updated.

**Comment by theanalyst** [comment]
Might be worth a backport @weiqiaomiao can you create a tracker issue as well

**Comment by joscollin** [comment]
@theanalyst Right, missed to mention.

@weiqiaomiao Update tracker URL in the commit.

**Comment by joscollin** [comment]
Jenkins retest this please

**Comment by weiqiaomiao** [comment]
@theanalyst @joscollin  OK～See the ticket：http://tracker.ceph.com/issues/21339

**Comment by yehudasa** [comment]
@weiqiaomiao @joscollin we hardly use try/catch anywhere, and I don't think this case merits its use. You can switch to something like `scope_guard` instead.

**Comment by theanalyst** [comment]
dropping the needs backport label, either way there is an issue to track, the backport field can be marked there.

**Comment by joscollin** [comment]
@yehudasa I understand. Yes, that's better.

**Comment by weiqiaomiao** [comment]
jenkins test this please

**Comment by weiqiaomiao** [comment]
@yehudasa @joscollin   I try to implement it with `make_scope_guard`. Could you take a look at it.Thanks!

**Pull request closed without merging**